src/platform_impl/linux/mod.rs

```diff
@@ -48,161 +71,192 @@ pub struct PlatformSpecificWindowBuilderAttributes {
 impl Default for PlatformSpecificWindowBuilderAttributes {
     fn default() -> Self {
         Self {
+            #[cfg(feature = "x11")]
             visual_infos: None,
             screen_id: None,
             resize_increments: None,
             base_size: None,
             class: None,
             override_redirect: false,
+            #[cfg(feature = "x11")]
             x11_window_types: vec![XWindowType::Normal],
             gtk_theme_variant: None,
             app_id: None,
         }
     }
 }
 
+#[cfg(feature = "x11")]
 lazy_static! {
     pub static ref X11_BACKEND: Mutex<Result<Arc<XConnection>, XNotSupported>> =
         Mutex::new(XConnection::new(Some(x_error_callback)).map(Arc::new));
 }
 
 #[derive(Debug, Clone)]
 pub enum OsError {
+    #[cfg(feature = "x11")]
     XError(XError),
     XMisc(&'static str),
 }
 
 impl fmt::Display for OsError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
         match self {
+            #[cfg(feature = "x11")]
             OsError::XError(e) => f.pad(&e.description),
             OsError::XMisc(e) => f.pad(e),
         }
     }
 }
 
 pub enum Window {
+    #[cfg(feature = "x11")]
     X(x11::Window),
+    #[cfg(feature = "wayland")]
     Wayland(wayland::Window),
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub enum WindowId {
+    #[cfg(feature = "x11")]
     X(x11::WindowId),
+    #[cfg(feature = "wayland")]
     Wayland(wayland::WindowId),
+    #[cfg(all(not(feature = "x11"), not(feature = "wayland")))]
+    Dummy,
 }
 
 impl WindowId {
     pub unsafe fn dummy() -> Self {
-        WindowId::Wayland(wayland::WindowId::dummy())
+        #[cfg(feature = "wayland")]
+        return WindowId::Wayland(wayland::WindowId::dummy());
+        #[cfg(all(not(feature = "wayland"), feature = "x11"))]
+        return WindowId::X(x11::WindowId::dummy());
+        #[cfg(all(not(feature = "x11"), not(feature = "wayland")))]
+        WindowId::Dummy
     }
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub enum DeviceId {
+    #[cfg(feature = "x11")]
     X(x11::DeviceId),
+    #[cfg(feature = "wayland")]
     Wayland(wayland::DeviceId),
+    #[cfg(all(not(feature = "x11"), not(feature = "wayland")))]
+    Dummy,
 }
 
 impl DeviceId {
     pub unsafe fn dummy() -> Self {
```

Comment:
If both x11 and wayland backends are disabled on unix, the compilation should be aborted I guess on build.rs stage, what do you think? That's the thing we're currently doing in alacritty, since we do expose those features as well.

There could be a potential in a future one could want `Dummy`, but the goal is to always get the best code now, so it seems like aborting when none are presented is a way to go.

Reply:
I disabled the compilation with a compile_error and removed the Dummy.